### PR TITLE
test: snapshot testing for unary stream executors

### PR DESCRIPTION
**Pull request by xxchan**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Introduce snapshot testing to make writing executor tests easier. See the doc in `src/stream/tests/it/snapshot.rs` for more details. 
 

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- ~~[ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).~~
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/stream/src/executor/over_window/snapshots/risingwave_stream__executor__over_window__eowc__tests__over_window.snap [hunk 1]
```diff
@@ -0,0 +1,61 @@
+---
+source: src/stream/src/executor/over_window/eowc.rs
+expression: "executor_snapshot(Box::new(async ||\n                create_executor(calls.clone(), store.clone()).await),\n        SnapshotEvent::parse(inputs)).await"
```
Comment (xxchan) [inline src/stream/src/executor/over_window/snapshots/risingwave_stream__executor__over_window__eowc__tests__over_window.snap:3]: Hide this `expression`

File: src/stream/src/executor/over_window/snapshots/risingwave_stream__executor__over_window__eowc__tests__over_window.snap
Comment (xxchan) [inline src/stream/src/executor/over_window/snapshots/risingwave_stream__executor__over_window__eowc__tests__over_window.snap:1]: Does this file location and file name make senes?

**Comment by xxchan** [comment]
PTAL and tell whether you like this thing.

**Comment by st1page** [comment]
how about splitting `src/stream/src/` and `src/stream/src/test`? In other words, the test in this PR will be in `src/stream/src/test/executor/over_window`. This arrangement of test maybe can not access some private field, but currently I think most of our test does not acquire access them at all.

**Comment by richardchien** [comment]
> Btw, I remember we have mentioned that we can move the unit test into a separate because our streaming executor's tests do not need access private field and methods. And after that we can depend on more utils such as frontend crate. But I can not find that comments or issue.

Not sure whether it's this comment: https://github.com/risingwavelabs/risingwave/issues/7881#issuecomment-1427930852

File: src/stream/src/executor/over_window/eowc.rs [hunk 2]
```diff
@@ -661,26 +622,19 @@ mod tests {
...
-        let chunk = over_window.expect_chunk().await;
-        assert_eq!(
-            chunk,
-            StreamChunk::from_pretty(
-                " I T  I   i  I
-                + 1 p1 100 10 26
-                + 1 p1 101 16 46"
+        insta::assert_snapshot!(
+            executor_snapshot(
+                Box::new(async || create_executor(calls.clone(), store.clone()).await),
```
Comment (BugenZhao) [inline src/stream/src/executor/over_window/eowc.rs:627]: Is async closure inevitable here?

**Review by BugenZhao — COMMENTED** [review]
I'm not sure whether there'll be many executors that can benefit from this. 🤔

- Some tests are directly manipulating the internal states.
- Some tests require more complicated control flows. (https://github.com/risingwavelabs/risingwave/blob/bd23cc0a2e907458696973791688f0827a0f1077/src/stream/src/executor/merge.rs#L553)

Actually, executor unit tests are not that "snapshot testing" to me. 🤔 The output is supposed to be manually derived by the developer and can be rarely changed. The ideas of simplifying these tests with DSL look good to me though.

**Comment by xxchan** [comment]
> I'm not sure whether there'll be many executors that can benefit from this. 

I also have similar concerns so created a demo to gain feedbacks :p 

I *guess* most executor tests are simple enough to be tested in this way. The only exceptions in my mind are:
- binary executors (i.e., JOIN)
- "control" executors

> Some tests are directly manipulating the internal states / require more complicated control flows

Yes, so we don't use snapshot testing for such more fine-grained tests.

> Actually, executor unit tests are not that "snapshot testing" to me. 🤔 The output is supposed to be manually derived by the developer and can be rarely changed. The ideas of simplifying these tests with DSL look good to me though.

The real motivation is that manually write these tests are too tiring and the test coverage is low!!! 😄 
- Mainly for adding new tests, instead of replacing current tests
- Generating outputs is good, and avoiding boilerplates is good. 

So we may hope to find a way to write the simple test pattern (overlapping input/output events) easier. The granularity is between e2e and complete unit tests. It's kind of 1-executor `slt` where we manipulate messages directly (I'm not sure whether it's possible and if so, whether it's easier to use `slt` for these tests).

File: src/stream/src/executor/over_window/eowc.rs (hunk 2 above)
Comment (xxchan) [reply src/stream/src/executor/over_window/eowc.rs:627]: I don't know (yet). Hope a professional to tell me. 🥵

Comment (BugenZhao) [reply src/stream/src/executor/over_window/eowc.rs]: What about a closure returning a `Future`? 🥵

**Comment by wangrunji0408** [comment]
I have thought about such test script and it looks pretty cool!
I'm just curious why it is called "snapshot" testing, instead of "executor" testing?

**Comment by richardchien** [comment]
> I'm just curious why it is called "snapshot" testing, instead of "executor" testing?

I agree with the name, it's comparing the output snapshot of a correct (committed) version of executor with latest one.

---

The idea looks great to me! Follow the same idea, can we also make e2e tests snapshot testing?

File: src/stream/src/executor/over_window/eowc.rs (hunk 2 above)
Comment (xxchan) [reply src/stream/src/executor/over_window/eowc.rs]: Oh damn

**Comment by xxchan** [comment]
> > I'm just curious why it is called "snapshot" testing, instead of "executor" testing?
> 
> I agree with the name, it's comparing the output snapshot of a correct (committed) version of executor with latest one.

Yes, that's true. To my understanding, "snapshot testing" is basically "file-based tests". The main idea is *just commit the entire output!*. 

It's also called "golden tests" (and the files are called golden files). Here's a blog introducing it https://www.cs.cornell.edu/~asampson/blog/turnt.html

> e2e tests snapshot testing (automatically generate output)?

I also have the same question but not sure. Sqllogictest and planner test both support *auto completion*, so to me they are already kind of "snapshot testing". The only difference is that they put input/output in the same file. Actually that's desired, but we can still do so (contain input in the generated output) for snapshots, which is already done in this PR.

Technically I think that's doable (and not hard). The only reason not to do it I can come up with is that current situation is good enough. (I think for sqllogictest that's true, but for planner tests we have some pain points and had some discussions about it https://github.com/risingwavelabs/risingwave/issues/8557)

**Review by soundOfDestiny — APPROVED** [review]
LGTM. thanks!

File: src/stream/src/executor/over_window/snapshots/risingwave_stream__executor__over_window__eowc__tests__over_window_aggregate.snap [hunk 3]
```diff
@@ -0,0 +1,23 @@
+---
+source: src/stream/src/executor/over_window/eowc.rs
+expression: "executor_snapshot(Box::new(async ||\n                create_executor(calls.clone(), store.clone()).await),\n        r###\"\n- barrier\n- !chunk |-\n    I T  I   i\n    + 1 p1 100 10\n    + 1 p1 101 16\n    + 4 p1 102 20\n\"###).await"
```
Comment (kwannoel) [inline src/stream/src/executor/over_window/snapshots/risingwave_stream__executor__over_window__eowc__tests__over_window_aggregate.snap:3]: Can we snapshot executor parameters here as well?

**Review by kwannoel — APPROVED** [review]
I think this will be very useful to improve test coverage. More importantly it can help us locate bug sources at the executor level. Currently I rely on e2e sql to test executor, but that is not ergonomic to use when we want test something specific for the executor.

Generally LGTM, further improvements can be made in subsequent PRs.

**Comment by xxchan** [comment]
Just found another lovely little tool [expect_test](https://github.com/rust-analyzer/expect-test). It's minimalistic and used by rust-analyzer. I find it might be enough for us.

A related question is do you prefer inline output string or separated output file?

Regarding the PROs and CONs of inline string:
  - pros
    - discoverability (no need to find the file)
    - all context at one place. no need to provide context for test files (e.g., `description`, `info` headers for `insta`)
    - better readability?
  - cons
    - ~~devs might be confused about how to update the result~~ (resolved by adding a note)
    - can also hurt readability?
    - larger possibility of merge conflict. This can be a larger problem for planner tests (https://github.com/risingwavelabs/risingwave/issues/8557#issuecomment-1473308653), but less a problem for executor tests, because there are fewer changes and the results are relatively stable.
    - The generated rust string (currently it's a yaml) contains indentation. Although it can be arbitrarily edited, I'm not sure whether somebody doesn't like it. (BTW, the input/output DSL in my demo is randomly built. Aren't there any comments about that? 🥵)

**Comment by xxchan** [comment]
About [expect_test](https://github.com/rust-analyzer/expect-test):

> It can be though of as a super-charged string literal, which can update itself. 
>
> Running the test with `UPDATE_EXPECT=1` env variable however would “magically” update the code

*It just works*. And doesn't need extra tooling. 😋

**Comment by xxchan** [comment]
Wait, so with `expect_test`, we can do a drop-in replacement:

```rust
let res = executor.next().await.unwrap().unwrap();
assert_eq!(
    *res.as_chunk().unwrap(),
    StreamChunk::from_pretty(
        " I I
        - 1 0"
    )
);
```

to

```rust
let expect = expect_test::expect![[r#"
    +---+---+---+
    | - | 1 | 0 |
    +---+---+---+"#]];
expect.assert_eq(&*res.as_chunk().unwrap().to_pretty_string());
```

So it might also enhance fine-grained UTs (as mentioned by @BugenZhao) with auto-completion support. 

BTW, I think the idea of DSL for simple cases is still valuable, as calling `next().await.unwrap().unwrap()` repeatedly is mostly just noise for reviewers and annoying for writers.

**Comment by xxchan** [comment]
This inspired me to come up with another idea: Just between "fully DSL" and "fully hand-written", a combination of both might be interesting: we "`run_until_pending`" and then print the output in the DSL form (which can be automatically generated). 

PTAL the latest push. With this I'm wondering whether writing input in DSL is necessary at all. 🤔

File: src/stream/src/executor/test_utils.rs [hunk 4]
```diff
@@ -545,3 +545,140 @@ pub mod top_n_executor {
...
+    /// ```
+    ///
+    /// Then just run the tests with env var `UPDATE_EXPECT=1`.
+    ///
+    /// ```shell
+    /// UPDATE_EXPECT=1 cargo nextest run -p risingwave_stream
+    /// # or
+    /// UPDATE_EXPECT=1 risedev test -p risingwave_stream
+    /// ```
+    pub async fn check_until_pending(
```
Comment (xxchan) [inline src/stream/src/executor/test_utils.rs:676]: Look at this!

File: src/stream/src/executor/top_n/top_n_plain.rs
Comment (xxchan) [inline src/stream/src/executor/top_n/top_n_plain.rs:1]: example of drop-in replacement with `check_until_pending`

File: src/stream/src/executor/hash_agg.rs
Comment (xxchan) [inline src/stream/src/executor/hash_agg.rs:1]: example of drop-in replacement with `check_until_pending`

File: src/stream/src/executor/over_window/eowc.rs
Comment (xxchan) [inline src/stream/src/executor/over_window/eowc.rs:1]: This file now contains many different styles for you to compare with ...

File: src/stream/src/executor/top_n/top_n_plain.rs [hunk 5]
```diff
@@ -863,53 +864,32 @@ mod tests {
...
+                      | - | 9 | 8 | 1 | 1004 |
+                      | + | 1 | 1 | 4 | 1001 |
+                      +---+---+---+---+------+
+                    - !chunk |-
+                      +---+---+---+---+------+
+                      | - | 5 | 1 | 4 | 1002 |
+                      | + | 1 | 0 | 2 | 1006 |
+                      +---+---+---+---+------+
+                    - !barrier 2
+                    - !barrier 7
```
Comment (xxchan) [inline src/stream/src/executor/top_n/top_n_plain.rs:889]: What's wrong 😅

**Comment by richardchien** [comment]
> This inspired me to come up with another idea: Just between "fully DSL" and "fully hand-written", a combination of both might be interesting: we "run_until_pending" and then print the output in the DSL form (which can be automatically generated).

The idea is interesting! But what if later someday we want to add mock state store that is not in-memory? Then the output is not available at the first pending.

**Comment by BugenZhao** [comment]
> calling next().await.unwrap().unwrap() repeatedly is mostly just noise for reviewers and annoying for writers.

IIRC, we have `ExecutorTestExt` for simplifying this. 🥰

File: src/stream/src/executor/test_utils.rs (hunk 4 above)
Comment (xxchan) [reply src/stream/src/executor/test_utils.rs]: > But what if later someday we want to add mock state store that is not in-memory? Then the output is not available at the first pending.

@richardchien  I wrote here we can use `check_n_steps` instead 🤪

**Comment by xxchan** [comment]
> > calling next().await.unwrap().unwrap() repeatedly is mostly just noise for reviewers and annoying for writers.
> 
> IIRC, we have `ExecutorTestExt` for simplifying this. 🥰

I know that. But what I like most is replacing multiple `expect` into one large one. How do you think

**Comment by xxchan** [comment]
My final decisions:
- all-in `expect_test`, because the workflow is simpler and it's good enough. 
- use inline output, because it might be better to review.
- keep both of the styles (code input and DSL script input), because I can't tell which is better.


Not decided:
- `check_until_pending` or `check_n_steps`

This PR is ready to merge. Final comments welcomed.

File: src/stream/tests/it/eowc.rs [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+- !barrier 2
+- recovery
+- !barrier 3
+- !chunk |2
+      I  T  I   i
+    + 10 p1 103 13
+    + 12 p2 202 28
+    + 13 p3 301 39
+- !barrier 4
+"###,
```
Comment (richardchien) [inline src/stream/tests/it/eowc.rs:116]: This looks a little bit ugly. Can we add indent here?

File: src/stream/tests/it/main.rs
Comment (richardchien) [inline src/stream/tests/it/main.rs:1]: Would you mind renaming `it` to `integration_tests` for the sake of clarity? `it` may be not so friendly for newcomers.

**Review by richardchien — COMMENTED** [review]
So is `expect![[]]` auto-updated on `risedev test`? What if developer forgets to run `risedev test` locally, will CI check that?

**Comment by xxchan** [comment]
> So is `expect![[]]` auto-updated on `risedev test`? What if developer forgets to run `risedev test` locally, will CI check that?

It's only updated if the envvar is set. It's just like a normal `assert_eq` otherwise (with diff in the error message). And there will also be a hint in the error message. See https://docs.rs/expect-test/latest/expect_test/ for more info

image

File: src/stream/tests/it/eowc.rs (hunk 6 above)
Comment (xxchan) [reply src/stream/tests/it/eowc.rs:116]: > Can we add indent here?

Yes

> This looks a little bit ugly.

I don't know. In theory this is a yaml, so somebody may argue that no indent make more sense. I have no preference 😇

**Comment by richardchien** [comment]
> > So is `expect![[]]` auto-updated on `risedev test`? What if developer forgets to run `risedev test` locally, will CI check that?
> 
> It's only updated if the envvar is set. It's just like a normal `assert_eq` otherwise (with diff in the error message). And there will also be a hint in the error message. See [docs.rs/expect-test/latest/expect_test](https://docs.rs/expect-test/latest/expect_test/) for more info

Will be nice to have sth like `risedev update-integration-tests`😁

**Comment by xxchan** [comment]
> Will be nice to have sth like `risedev update-integration-tests`😁

I don't want to add a script for such simple task and feel env var is good enough. But I don't object to adding it neither. 🤪

**Pull request merged**